### ISSUE TITLE
Extrapolation experiments

### ISSUE THREAD

**Issue by drbwa** [arise-insights/arise-predictions#54]
Our existing approach to finding the best learning algorithm and hyperparameters for a given data set results in models with good predictive performance for input features whose values are within the ranges of values encountered during training. However, when trying to predict for ranges of input feature values that exceed those seen during training, our models' performance degrades significantly. Often, the best models for a data set are tree-based models (e.g., Random Forest, XGBoost), which have known shortcomings for extrapolation. We want to improve the performance of our models for extrapolation. 

Consider predictions of throughput (e.g., tokens per second) for different input token lengths. Running experiments for a large set of large input token lengths can become expensive. It would be preferable, if we could obtain ample experimental data for smaller input token lengths (e.g., 1 - 1,024 under various settings of other input features) and limit the number of experimental data points for higher data points to a minimum (e.g., one or two experiments at input token lengths of 16K, one at 32K, one at 64K).

## What we have

- Code to pick the best learning algorithm and hyper-parameters for a given data set and target variable.
- Code to split data set into train-test and extrapolation test set based on a single input feature and thresholds over its values.
- Code to test extrapolation of best model.
- Using MAPE, NRMSE (by range) as metrics.

## What we have tried

Tried using a stacked regressor to combine the predictions of the best-performing linear and non-linear models into a meta-learner. Tried a number of algorithms as the meta-learned.

Improve performance significantly, but not within useful range (e.g., improve MAPE from about 80% to 40%).

### Transfer learning 

1. Build base model trained on training data set as we do today. This base model performs well within the range of values encountered during training.
2. Experiment with transfer learning on a small number of data points from the region we want to extrapolate to. 

- How many data points in the extrapolation range are sufficient to get 'good' performance?
- What is the effect on model performance of adding intermediate training data points in the extrapolation range? Or is it sufficient to use data points at the extreme ends of the extrapolation range?

Baseline assessment

- Train base model on ample data with input token lengths up to some limit (e.g., 2K).
- Measure performance of base model for predictions within range.
- Measure performance of base model for predictions in extrapolation range.

Extrapolation experiments:

1. Perform transfer learning with one data points only at the extreme end of the extrapolation range (e.g., input token length of 100K). 
2. Measure performance of transfer-learned model at various points in the extrapolation range (e.g., 4K, 8K, 16K, etc.)
3. Repeat (1.) and (2.) with more data points at extreme end of extrapolation range.
4. Repeat (1.) and (2.) using varying amounts of intermediate training points from the extrapolation range.

Extensions:

- Extend to multiple extrapolation features.
- Can we learn to predict how much data and in what ranges are needed?
